src/main/java/ca/uhn/fhir/jpa/starter/interceptor/CustomConsentService.java

```diff
@@ -0,0 +1,224 @@
+      return true;
+    }
+    return false;
+  }
+
+  private String getPatientClaim(RequestDetails theRequestDetails) {
+    return OAuth2Helper.getClaimAsString(theRequestDetails, "patient");
+  }
+
+  private ConsentOutcome startPostOperation(RequestDetails theRequestDetails, String patientId) {
+    boolean proceed = isResourceForPatient(theRequestDetails.getResource(), patientId);
+    if (logger.isDebugEnabled()) {
+      String decision = proceed ? "Allowing" : "Denying";
+      logger.debug("{} POST operation", decision);
+    }
+    return proceed ? ConsentOutcome.PROCEED : ConsentOutcome.REJECT;
+  }
+
+  private ConsentOutcome startPutOperation(RequestDetails theRequestDetails, String patientId) {
+    // Check for the persistent resource first to ensure that a 404 error is returned when trying
+    // to update any resource that is NOT for the patient in the token. Otherwise, this code would
+    // cause a 403; which would "leak" the knowledge (of the id) of a resource that the patient
+    // is not allowed access to.
+    boolean proceed = isResourceForPatient(getPersistentResource(theRequestDetails), patientId)
+      && isResourceForPatient(theRequestDetails.getResource(), patientId);
+    if (logger.isDebugEnabled()) {
+      String decision = proceed ? "Allowing" : "Denying";
+      logger.debug("{} PUT operation", decision);
+    }
+    return proceed ? ConsentOutcome.PROCEED : ConsentOutcome.REJECT;
+  }
+
+  private ConsentOutcome startPatchOperation(RequestDetails theRequestDetails, String patientId) {
+    // Check for the persistent resource first to ensure that a 404 error is returned when trying
+    // to update any resource that is NOT for the patient in the token. Otherwise, this code would
+    // cause a 403; which would "leak" the knowledge (of the id) of a resource that the patient
+    // is not allowed access to.
+    boolean proceed = isResourceForPatient(getPersistentResource(theRequestDetails), patientId)
+      && isPatchRequestBodyValid(theRequestDetails, patientId);
+    if (logger.isDebugEnabled()) {
+      String decision = proceed ? "Allowing" : "Denying";
+      logger.debug("{} PATCH operation", decision);
+    }
+    return proceed ? ConsentOutcome.PROCEED : ConsentOutcome.REJECT;
+  }
+
+  private ConsentOutcome startDeleteOperation(RequestDetails theRequestDetails, String patientId) {
+    boolean proceed = isResourceForPatient(getPersistentResource(theRequestDetails), patientId);
+    if (logger.isDebugEnabled()) {
+      String decision = proceed ? "Allowing" : "Denying";
+      logger.debug("{} DELETE operation", decision);
+    }
+    return proceed ? ConsentOutcome.PROCEED : ConsentOutcome.REJECT;
+  }
+
+  private IBaseResource getPersistentResource(RequestDetails theRequestDetails) {
+    String[] parts = theRequestDetails.getRequestPath().split("/");
+    return daoRegistry.getResourceDao(parts[0]).read(new IdType(parts[1]), theRequestDetails);
```

Comment:
This code needs to handle `ca.uhn.fhir.rest.server.exceptions.ResourceNotFoundException` .
In case of PUT opeartion if the patient2 token tries to update patient1 task it is returning 404 instead of 403 Rejected by consent service

Reply:
It should return a 404 - not a 403. It returning a 403 before is a bug.

Returning anything other than a 404 would "leak" the knowledge of a resource that the patient doesn't have access to.